## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :x: [more...](#7a0b48d4-4639-4126-809f-4b561027477c) |
| mint-fs.sh | :x: [more...](#acd4d9da-18b8-4024-8d7a-d36588f132ab) |
| mint-gateway-nas.sh | :x: [more...](#35aad140-fe8d-49c7-b502-78f29b163a21) |
<a name="acd4d9da-18b8-4024-8d7a-d36588f132ab"></a>
### 7387-3f1228c/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:32323
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d722d43d626a:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 41 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 46 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 14 seconds
(8/14) Running minio-dotnet tests ... done in 22 seconds
(9/14) Running minio-go tests ... done in 12 seconds
(10/14) Running minio-java tests ... FAILED in 32 seconds
{
  "name": "minio-java",
  "function": "copyObject(String bucketName, String objectName, String destBucketName, CopyConditions copyConditions)",
  "args": "CopyConditions: invalidETag",
  "duration": 607,
  "status": "FAIL",
  "error": "java.net.SocketException: Connection reset >>> [java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:115), java.net.SocketOutputStream.write(SocketOutputStream.java:155), okio.Okio$1.write(Okio.java:78), okio.AsyncTimeout$1.write(AsyncTimeout.java:179), okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.java:171), okio.RealBufferedSink.write(RealBufferedSink.java:41), okhttp3.internal.http1.Http1Codec$FixedLengthSink.write(Http1Codec.java:286), okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.java:171), okio.RealBufferedSink.write(RealBufferedSink.java:109), io.minio.HttpRequestBody.writeTo(HttpRequestBody.java:74), okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:62), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:185), okhttp3.RealCall.execute(RealCall.java:69), io.minio.MinioClient.executeReq(MinioClient.java:1097), io.minio.MinioClient.execute(MinioClient.java:1056), io.minio.MinioClient.executePut(MinioClient.java:1399), io.minio.MinioClient.executePut(MinioClient.java:1421), io.minio.MinioClient.putObject(MinioClient.java:3994), io.minio.MinioClient.putObject(MinioClient.java:4037), io.minio.MinioClient.putObject(MinioClient.java:3578), FunctionalTest.copyObject_test2(FunctionalTest.java:2054), FunctionalTest.runTests(FunctionalTest.java:3040), FunctionalTest.main(FunctionalTest.java:3154)]"
}

Executed 9 out of 14 tests successfully.
```
<a name="7a0b48d4-4639-4126-809f-4b561027477c"></a>
### 7387-3f1228c/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:30812
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 63a457764283:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 44 seconds
(4/14) Running aws-sdk-ruby tests ... done in 4 seconds
(5/14) Running awscli tests ... done in 1 minutes and 53 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 32 seconds
(8/14) Running minio-dotnet tests ... done in 43 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 13 seconds
(10/14) Running minio-java tests ... done in 2 minutes and 55 seconds
(11/14) Running minio-js tests ... done in 40 seconds
(12/14) Running minio-py tests ... FAILED in 3 minutes and 30 seconds
{
  "message": "AccessDenied: message: Access Denied",
  "args": {
    "object_name": "aa964731-bc60-4c32-82ac-86c4149e8526",
    "bucket_name": "minio-py-test-af77178f-6b06-4892-b91f-b161b193d39d"
  },
  "status": "FAIL",
  "duration": 123,
  "name": "minio-py:test_presigned_get_object_expiry_5sec",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1220, in test_presigned_get_object_expiry_5sec\n    object_name).get_exception()\nminio.error.AccessDenied: AccessDenied: message: Access Denied\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1885, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1228, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: AccessDenied: message: Access Denied\n",
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)"
}

Executed 11 out of 14 tests successfully.
```
<a name="35aad140-fe8d-49c7-b502-78f29b163a21"></a>
### 7387-3f1228c/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:32110
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 8c18a2da9f54:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 1 minutes and 47 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 22 seconds
(8/14) Running minio-dotnet tests ... done in 25 seconds
(9/14) Running minio-go tests ... done in 33 seconds
(10/14) Running minio-java tests ... done in 1 minutes and 33 seconds
(11/14) Running minio-js tests ... done in 25 seconds
(12/14) Running minio-py tests ... FAILED in 2 minutes and 37 seconds
{
  "duration": 50,
  "name": "minio-py:test_presigned_get_object_expiry_5sec",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1220, in test_presigned_get_object_expiry_5sec\n    object_name).get_exception()\nminio.error.AccessDenied: AccessDenied: message: Access Denied\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1885, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1228, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: AccessDenied: message: Access Denied\n",
  "message": "AccessDenied: message: Access Denied",
  "args": {
    "bucket_name": "minio-py-test-85b0ed11-1e91-4fb4-a7ab-2f7d6967ea1c",
    "object_name": "ba6c515e-ee8d-4a56-92bd-f834ff8df8e6"
  },
  "status": "FAIL",
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)"
}

Executed 11 out of 14 tests successfully.
```

--- a/cmd/api-errors.go
+++ b/cmd/api-errors.go
@@ -180,6 +180,7 @@ const (
 	// Minio extended errors.
 	ErrReadQuorum
 	ErrWriteQuorum
+	ErrParentIsObject
 	ErrStorageFull
 	ErrRequestBodyParse
 	ErrObjectExistsAsDirectory
@@ -857,6 +858,11 @@ var errorCodes = errorCodeMap{
 		Code:           "XMinioStorageFull",
 		Description:    "Storage backend has reached its minimum free disk threshold. Please delete a few objects to proceed.",
 		HTTPStatusCode: http.StatusInsufficientStorage,
+	},
+	ErrParentIsObject: {
+		Code:           "XMinioParentIsObject",
+		Description:    "Object-prefix is already an object, please choose a different object-prefix name.",
+		HTTPStatusCode: http.StatusBadRequest,
 	},
 	ErrRequestBodyParse: {
 		Code:           "XMinioRequestBodyParse",
@@ -1563,6 +1569,8 @@ func toAPIErrorCode(ctx context.Context, err error) (apiErr APIErrorCode) {
 		apiErr = ErrObjectExistsAsDirectory
 	case PrefixAccessDenied:
 		apiErr = ErrAccessDenied
+	case ParentIsObject:
+		apiErr = ErrParentIsObject
 	case BucketNameInvalid:
 		apiErr = ErrInvalidBucketName
 	case BucketNotFound:

--- a/cmd/disk-cache-fs.go
+++ b/cmd/disk-cache-fs.go
@@ -332,7 +332,7 @@ func (cfs *cacheFSObjects) PutObject(ctx context.Context, bucket string, object 
 	if isObjectDir(object, data.Size()) {
 		// Check if an object is present as one of the parent dir.
 		if fs.parentDirIsObject(ctx, bucket, path.Dir(object)) {
-			return ObjectInfo{}, toObjectErr(errFileAccessDenied, bucket, object)
+			return ObjectInfo{}, toObjectErr(errFileParentIsFile, bucket, object)
 		}
 		if err = mkdirAll(pathJoin(fs.fsPath, bucket, object), 0777); err != nil {
 			return ObjectInfo{}, toObjectErr(err, bucket, object)
@@ -350,7 +350,7 @@ func (cfs *cacheFSObjects) PutObject(ctx context.Context, bucket string, object 
 
 	// Check if an object is present as one of the parent dir.
 	if fs.parentDirIsObject(ctx, bucket, path.Dir(object)) {
-		return ObjectInfo{}, toObjectErr(errFileAccessDenied, bucket, object)
+		return ObjectInfo{}, toObjectErr(errFileParentIsFile, bucket, object)
 	}
 
 	// Validate input data size and it can never be less than zero.

--- a/cmd/fs-v1-multipart.go
+++ b/cmd/fs-v1-multipart.go
@@ -487,7 +487,7 @@ func (fs *FSObjects) CompleteMultipartUpload(ctx context.Context, bucket string,
 
 	// Check if an object is present as one of the parent dir.
 	if fs.parentDirIsObject(ctx, bucket, pathutil.Dir(object)) {
-		return oi, toObjectErr(errFileAccessDenied, bucket, object)
+		return oi, toObjectErr(errFileParentIsFile, bucket, object)
 	}
 
 	if _, err := fs.statBucketDir(ctx, bucket); err != nil {

--- a/cmd/fs-v1.go
+++ b/cmd/fs-v1.go
@@ -839,8 +839,7 @@ func (fs *FSObjects) putObject(ctx context.Context, bucket string, object string
 	if isObjectDir(object, data.Size()) {
 		// Check if an object is present as one of the parent dir.
 		if fs.parentDirIsObject(ctx, bucket, path.Dir(object)) {
-			logger.LogIf(ctx, errFileAccessDenied)
-			return ObjectInfo{}, toObjectErr(errFileAccessDenied, bucket, object)
+			return ObjectInfo{}, toObjectErr(errFileParentIsFile, bucket, object)
 		}
 		if err = mkdirAll(pathJoin(fs.fsPath, bucket, object), 0777); err != nil {
 			logger.LogIf(ctx, err)
@@ -859,8 +858,7 @@ func (fs *FSObjects) putObject(ctx context.Context, bucket string, object string
 
 	// Check if an object is present as one of the parent dir.
 	if fs.parentDirIsObject(ctx, bucket, path.Dir(object)) {
-		logger.LogIf(ctx, errFileAccessDenied)
-		return ObjectInfo{}, toObjectErr(errFileAccessDenied, bucket, object)
+		return ObjectInfo{}, toObjectErr(errFileParentIsFile, bucket, object)
 	}
 
 	// Validate input data size and it can never be less than zero.

--- a/cmd/fs-v1_test.go
+++ b/cmd/fs-v1_test.go
@@ -230,8 +230,8 @@ func TestFSPutObject(t *testing.T) {
 	if err == nil {
 		t.Fatal("Unexpected should fail here, backend corruption occurred")
 	}
-	if nerr, ok := err.(PrefixAccessDenied); !ok {
-		t.Fatalf("Expected PrefixAccessDenied, got %#v", err)
+	if nerr, ok := err.(ParentIsObject); !ok {
+		t.Fatalf("Expected ParentIsObject, got %#v", err)
 	} else {
 		if nerr.Bucket != "bucket" {
 			t.Fatalf("Expected 'bucket', got %s", nerr.Bucket)
@@ -245,8 +245,8 @@ func TestFSPutObject(t *testing.T) {
 	if err == nil {
 		t.Fatal("Unexpected should fail here, backned corruption occurred")
 	}
-	if nerr, ok := err.(PrefixAccessDenied); !ok {
-		t.Fatalf("Expected PrefixAccessDenied, got %#v", err)
+	if nerr, ok := err.(ParentIsObject); !ok {
+		t.Fatalf("Expected ParentIsObject, got %#v", err)
 	} else {
 		if nerr.Bucket != "bucket" {
 			t.Fatalf("Expected 'bucket', got %s", nerr.Bucket)

--- a/cmd/object-api-errors.go
+++ b/cmd/object-api-errors.go
@@ -19,6 +19,7 @@ package cmd
 import (
 	"fmt"
 	"io"
+	"path"
 )
 
 // Converts underlying storage error. Convenience function written to
@@ -43,6 +44,13 @@ func toObjectErr(err error, params ...string) error {
 	case errFileAccessDenied:
 		if len(params) >= 2 {
 			err = PrefixAccessDenied{
+				Bucket: params[0],
+				Object: params[1],
+			}
+		}
+	case errFileParentIsFile:
+		if len(params) >= 2 {
+			err = ParentIsObject{
 				Bucket: params[0],
 				Object: params[1],
 			}
@@ -180,6 +188,13 @@ type PrefixAccessDenied GenericError
 
 func (e PrefixAccessDenied) Error() string {
 	return "Prefix access is denied: " + e.Bucket + "/" + e.Object
+}
+
+// ParentIsObject object access is denied.
+type ParentIsObject GenericError
+
+func (e ParentIsObject) Error() string {
+	return "Parent is object " + e.Bucket + "/" + path.Dir(e.Object)
 }
 
 // BucketExists bucket exists.

--- a/cmd/os-reliable.go
+++ b/cmd/os-reliable.go
@@ -142,6 +142,10 @@ func renameAll(srcFilePath, dstFilePath string) (err error) {
 			return fmt.Errorf("%s (%s)->(%s)", errCrossDeviceLink, srcFilePath, dstFilePath)
 		case os.IsNotExist(err):
 			return errFileNotFound
+		case os.IsExist(err):
+			// This is returned only when destination is a directory and we
+			// are attempting a rename from file to directory.
+			return errIsNotRegular
 		default:
 			return err
 		}

--- a/cmd/storage-errors.go
+++ b/cmd/storage-errors.go
@@ -66,8 +66,11 @@ var errVolumeNotEmpty = errors.New("volume is not empty")
 // errVolumeAccessDenied - cannot access volume, insufficient permissions.
 var errVolumeAccessDenied = errors.New("volume access denied")
 
-// errVolumeAccessDenied - cannot access file, insufficient permissions.
+// errFileAccessDenied - cannot access file, insufficient permissions.
 var errFileAccessDenied = errors.New("file access denied")
+
+// errFileParentIsFile - cannot have overlapping objects, parent is already a file.
+var errFileParentIsFile = errors.New("parent is a file")
 
 // errBitrotHashAlgoInvalid - the algo for bit-rot hash
 // verification is empty or invalid.

--- a/cmd/xl-v1-multipart.go
+++ b/cmd/xl-v1-multipart.go
@@ -624,7 +624,7 @@ func (xl xlObjects) CompleteMultipartUpload(ctx context.Context, bucket string, 
 	// Check if an object is present as one of the parent dir.
 	// -- FIXME. (needs a new kind of lock).
 	if xl.parentDirIsObject(ctx, bucket, path.Dir(object)) {
-		return oi, toObjectErr(errFileAccessDenied, bucket, object)
+		return oi, toObjectErr(errFileParentIsFile, bucket, object)
 	}
 
 	// Calculate s3 compatible md5sum for complete multipart.

--- a/cmd/xl-v1-object.go
+++ b/cmd/xl-v1-object.go
@@ -577,7 +577,7 @@ func (xl xlObjects) putObject(ctx context.Context, bucket string, object string,
 		// -- FIXME. (needs a new kind of lock).
 		// -- FIXME (this also causes performance issue when disks are down).
 		if xl.parentDirIsObject(ctx, bucket, path.Dir(object)) {
-			return ObjectInfo{}, toObjectErr(errFileAccessDenied, bucket, object)
+			return ObjectInfo{}, toObjectErr(errFileParentIsFile, bucket, object)
 		}
 
 		if err = xl.putObjectDir(ctx, minioMetaTmpBucket, tempObj, writeQuorum); err != nil {
@@ -608,7 +608,7 @@ func (xl xlObjects) putObject(ctx context.Context, bucket string, object string,
 	// -- FIXME. (needs a new kind of lock).
 	// -- FIXME (this also causes performance issue when disks are down).
 	if xl.parentDirIsObject(ctx, bucket, path.Dir(object)) {
-		return ObjectInfo{}, toObjectErr(errFileAccessDenied, bucket, object)
+		return ObjectInfo{}, toObjectErr(errFileParentIsFile, bucket, object)
 	}
 
 	// Limit the reader to its provided size if specified.

